### PR TITLE
[CALCITE-841] In ProjectToWindowRule, RexOver will be put in the same…

### PR DESCRIPTION
**Pull request by hsuanyi**
… cohort if the following conditions are satisfied

(1). They have the same RexWindow
(2). They are not dependent on each other

### CODE REVIEW

**Comment by julianhyde** [comment]
Please don't put spaces at ends of lines: "... \n". You end up with spaces at ends of lines in RelOptRulesTest.xml, and then later there are spurious diffs when someone's IDE removes the spaces.

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectToWindowRule.java [hunk 1]
```diff
@@ -239,11 +250,131 @@ protected RelNode makeRel(
...
+
+    private List<Integer> getRank(final List<RexNode> exprs) {
+      final DirectedGraph<Integer, DefaultEdge> graph =
+          createGraphFromExpression(exprs);
+      TopologicalOrderIterator<Integer, DefaultEdge> iter =
+          new TopologicalOrderIterator<Integer, DefaultEdge>(graph);
+      final Integer[] rankArr = new Integer[exprs.size()];
+      int rank = 0;
+      while (iter.hasNext()) {
+        rankArr[iter.next()] = rank;
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectToWindowRule.java]: don't you mean 'rank++'?

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectToWindowRule.java [hunk 2]
```diff
@@ -239,11 +250,131 @@ protected RelNode makeRel(
...
+        final List<RexNode> exprs) {
+      final DirectedGraph<Integer, DefaultEdge> graph =
+          DefaultDirectedGraph.create();
+      for (int i = 0; i < exprs.size(); i++) {
+        graph.addVertex(i);
+      }
+
+      for (int i = 0; i < exprs.size(); i++) {
+        final RexNode expr = exprs.get(i);
+        final Set<Integer> targets = Collections.singleton(i);
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectToWindowRule.java]: why create a singleton set then iterate over it?

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectToWindowRule.java [hunk 3]
```diff
@@ -239,11 +250,131 @@ protected RelNode makeRel(
...
+        final List<Integer> rank,
+        final int ordinal1,
+        final int ordinal2) {
+      if (rank.get(ordinal2) > rank.get(ordinal1)) {
+        return isDependent(graph, rank, ordinal2, ordinal1);
+      }
+
+      // Check if the expression in ordinal1
+      // could depend on expression in ordinal2 by Depth-First-Search
+      final Stack<Integer> dfs = new Stack<>();
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectToWindowRule.java]: ArrayDeque is better than Stack. E.g. it is not synchronized

File: core/src/main/java/org/apache/calcite/rel/rules/ProjectToWindowRule.java [hunk 4]
```diff
@@ -239,11 +250,131 @@ protected RelNode makeRel(
...
+          }
+        }
+      }
+
+      return false;
+    }
+
+    private List<Integer> getRank(final List<RexNode> exprs) {
+      final DirectedGraph<Integer, DefaultEdge> graph =
+          createGraphFromExpression(exprs);
```
Comment (julianhyde) [inline core/src/main/java/org/apache/calcite/rel/rules/ProjectToWindowRule.java]: could avoid building graph again if we pass in the graph we just built

**Comment by julianhyde** [comment]
These are all minor nits. I have fixed them and rebased and will commit when the tests pass.

**Comment by hsuanyi** [comment]
Thanks Julian!

**Pull request closed without merging**